source/core/MicroBitFiber.cpp

```diff
@@ -65,6 +67,44 @@ static EventModel *messageBus = NULL;
 // Array of components which are iterated during idle thread execution.
 static MicroBitComponent* idleThreadComponents[MICROBIT_IDLE_COMPONENTS];
 
+static void get_fibers_from(Fiber ***dest, int *sum, Fiber *queue)
+{
+    if (queue && queue->prev)
+        microbit_panic(MICROBIT_HEAP_ERROR);
+    while (queue) {
+        if (*dest)
+            *(*dest)++ = queue;
+        (*sum)++;
+        queue = queue->next;
+    }
+}
+
+/**
+  * Return all current fibers.
+  *
+  * @param dest If non-null, it points to an array of pointers to fibers to store results in.
+  *
+  * @return the number of fibers (potentially) stored
+  */
+int list_fibers(Fiber **dest)
+{
+    int sum = 0;
+
+    // interrupts might move fibers between queues, but should not create new ones
+    __disable_irq();
+    get_fibers_from(&dest, &sum, runQueue);
+    get_fibers_from(&dest, &sum, sleepQueue);
+    get_fibers_from(&dest, &sum, waitQueue);
```

Comment:
Looks ok, but don't forget we introduced a mutex semaphore as part of js-event-semantics... Each mutex has its own wait queue associated with it, so you could perhaps miss extracting those fiber context with this function? Would that cause a problem for your application of this?

https://github.com/lancaster-university/microbit-dal/blob/js-event-semantics/source/core/MicroBitFiber.cpp#L944